src/network/networkd-dhcp6.c

```diff
@@ -842,3 +963,17 @@ static int dhcp6_prefix_remove_all(Manager *m, Link *link) {
 
         return 0;
 }
+
+static bool dhcp6_link_has_dhcpv6_prefix(Link *link) {
+        Iterator i;
+        Link *l;
+
+        assert(link);
+        assert(link->manager);
+
+        HASHMAP_FOREACH(l, link->manager->dhcp6_prefixes, i) {
+                if (link == l) return true;
+        }
+
+        return false;
```

Comment:
hmm, O(n). Is that going to be a problem? i.e how many prefixes do we expect here? The thing is currently a hashtable keyed by address afaics. should there be another hashtable, keyed by link?

Reply:
I was thinking about that as well for some time during the initial implementation.

The situation is that most people usually just need <10 prefixes in most "simple" scenarios. So this isn't a big deal. Some people might have a tens or even a hundred interfaces that need prefixes. Sounds reasonable, right?

The delegated prefixes usually allow us to serve thousands of interfaces with prefixes. A usual home-router setup (i.e. like from a Fritz!Box) probably just need 2-3 prefixes ("LAN", "Guests", and IoT or similar) so for those it would probably not pay off having another data structure. 

Since usually you have address space for 2^8 or 2^16 prefixes being able to handle these amounts of prefixes would be desirable. I wondered how networkd behaves with a small number of many interfaces that require prefixes. So I just wrote a quick test where I created N interfaces, each of them should get a prefix allocated. When growing N the time spent on configuring interfaces should also grow to some degree but otherwise it should work.

I started out with N=200 as that seemed to be a reasonable amount of prefixes for a non-trivial setup. The allocation code seems to run just fine and the time spent on calculating the prefixes (doing the hashtable iterations etc..) seems to be okayish. The problem is just that networkd is unable to configure that many interface addresses at the same time. I saw a bunch of Assertions and timeouts scrolling by and in the end none of the dummy interfaces had addresses from the prefixes configured. So it might be that at ~200 interfaces (and thus ~200 prefixes) the time spent on the hashtable are negligible given that the configuration doesn't even work for that amount of interfaces. 

Here is the log of my small test setup where I spawn two VMs. One of them hands out /48 prefixes and the other one ("router") is the node that runs networkd and tries to distribute them across the 200 (dummy) interfaces: https://gist.githubusercontent.com/andir/616f937189d176da86dbe7aa0f65217e/raw/918c9926f66c5788d9de2795dee5200331daf8c5/l

Looking at the log the two virtio interfaces did get their addresses assigned - those had been created by the kernel. All those dummy interfaces that have been created by networkd ran into the timeout. Not sure how all of this correlates. Maybe I just have to test with a large amount of "real" (virtual) interfaces.

Reply:
Oh, that sounds bad. It would be great to figure out where this breaks down. 200 interfaces is not crazy at all, this should work.

Reply:
This looks very similar to https://github.com/systemd/systemd/issues/14417#issuecomment-629529020 so I commented there.